adafruit_rfm/rfm_common.py

```diff
@@ -0,0 +1,545 @@
+# SPDX-FileCopyrightText: 2024 Jerry Needell for Adafruit Industries
+#
+# SPDX-License-Identifier: MIT
+
+"""
+
+* Author(s): Jerry Needell
+"""
+
+import asyncio
+import random
+import time
+
+from adafruit_bus_device import spi_device
+
+try:
+    from typing import Callable, Optional, Type
+
+    import busio
+    import digitalio
+    from circuitpython_typing import ReadableBuffer, WriteableBuffer
+
+except ImportError:
+    pass
+
+from micropython import const
+
+HAS_SUPERVISOR = False
+
+try:
+    import supervisor
+
+    if hasattr(supervisor, "ticks_ms"):
+        HAS_SUPERVISOR = True
+except ImportError:
+    pass
+
+
+__version__ = "0.0.0+auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_RFM.git"
+
+
+# RadioHead specific compatibility constants.
+_RH_BROADCAST_ADDRESS = const(0xFF)
+
+# The acknowledgement bit in the FLAGS
+# The top 4 bits of the flags are reserved for RadioHead. The lower 4 bits are reserved
+# for application layer use.
+_RH_FLAGS_ACK = const(0x80)
+_RH_FLAGS_RETRY = const(0x40)
+
+
+# supervisor.ticks_ms() contants
+_TICKS_PERIOD = const(1 << 29)
+_TICKS_MAX = const(_TICKS_PERIOD - 1)
+_TICKS_HALFPERIOD = const(_TICKS_PERIOD // 2)
+
+
+def ticks_diff(ticks1: int, ticks2: int) -> int:
+    """Compute the signed difference between two ticks values
+    assuming that they are within 2**28 ticks
+    """
+    diff = (ticks1 - ticks2) & _TICKS_MAX
+    diff = ((diff + _TICKS_HALFPERIOD) & _TICKS_MAX) - _TICKS_HALFPERIOD
+    return diff
+
+
+def asyncio_to_blocking(function):
+    """run async function as normal blocking function"""
+
+    def blocking_function(self, *args, **kwargs):
+        return asyncio.run(function(self, *args, **kwargs))
+
+    return blocking_function
+
+
+async def asyncio_check_timeout(flag: Callable, limit: float, timeout_poll: float) -> bool:
+    """test for timeout waiting for specified flag"""
+    timed_out = False
+    if HAS_SUPERVISOR:
+        start = supervisor.ticks_ms()
+        while not timed_out and not flag():
+            if ticks_diff(supervisor.ticks_ms(), start) >= limit * 1000:
+                timed_out = True
+            await asyncio.sleep(timeout_poll)
+    else:
+        start = time.monotonic()
+        while not timed_out and not flag():
+            if time.monotonic() - start >= limit:
+                timed_out = True
+            await asyncio.sleep(timeout_poll)
+
+    return timed_out
+
+
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-many-nested-blocks
+class RFMSPI:
+    """Base class for SPI type devices"""
+
+    class RegisterBits:
```

Comment:
Could you expand the separate register library instead to do this?

Reply:
You mean that I should try to change to using the adafruit_register library -- I'll take a look.

Reply:
Yes, please.

Reply:
I'll be happy to take this on.  It will require adding the SPI code to the adafruit_register library and a minor addition to the circuitPython_typing library to add SPIDevice.
Would it make sense to go ahead and create an initial release of the adafruit_rfm library with the existing register bit handling then update it to use the updated adafruit_register library when I get it working and tested.
If you prefer to wait until I have adafruit_register updated, that is fine with me.

Reply:
A follow up PR is fine.